EIPS/eip-1066.md

```diff
@@ -0,0 +1,314 @@
+
+As is the case with HTTP, having a standard set of known codes has many benefits for developers.
+They remove friction from needing to develop your own schemes for every contract,
+makes inter-contract automation easier, and makes it easier to broadly understand
+which of the finite states your request produced. Importantly, it makes it much easier
+to distinguish between expected errors states, and truly exceptional conditions
+that require halting execution.
+
+## Motivation
+
+### Autonomy
+
+Smart contracts are largely intended to be autonomous. While each contract may
+define a specific interface, having a common set of semantic codes can help
+developers write code that can react appropriately to various situations.
+
+### Semantic Density
+
+HTTP status codes are widely used for this purpose. BEAM languages use atoms
+and tagged tuples to signify much the same information. Both provide a lot of
+information both to the programmer (debugging for instance), and to the program
+that needs to decide what to do next.
+
+ESCs convey a much richer set of information than Booleans,
+and are able to be reacted to autonomously unlike arbitrary strings.
+
+### User Feedback
+
+Since status codes are finite and known in advance, we can provide global,
+human-readable sets of status messages. These may also be translated into any language,
+differing levels of technical detail, added as `revert` messages, natspecs, and so on.
+
+We also see a desire for this [in transactions](http://eips.ethereum.org/EIPS/eip-658),
+and there's no reason that ESCs couldn't be used by the EVM itself.
+
+### More than Pass/Fail
+
+While clearly related, status codes are complementary to "revert with reason".
+ESCs are not limited to rolling back the transaction, and may represent known error states
+without halting execution. They may also represent off-chain conditions,
+supply a string to revert, signal time delays, and more.
+
+## Specification
+
+### Format
+
+Codes are returned as the first value of potentially multiple return values.
+
+```solidity
+// Status only
+
+function isInt(uint num) public pure returns (byte status) {
+    return hex"01";
+}
+
+// Status and value
+
+uint8 private counter;
+
+function safeIncrement(uint8 interval) public returns (byte status, uint8 newCounter) {
+    uint8 updated = counter + interval;
+
+    if (updated >= counter) {
+        counter = updated;
+        return (hex"01", updated);
+    } else {
+        return (hex"00", counter);
+    }
+}
+```
+
+In the rare case that there a multiple codes required to express an idea,
+they should be organized in asending order.
+
+### Code Table
+
+| X. Low Nibble                     | 0. Generic              | 10. Permission                | 20. Find/Match/&c           | 30. Negotiation / Offers         | 40. Availability                 | 50. | 60. | 70. | 80. | 90. | A0. | B0. | C0. | D0. | E0. Cryptography & Tokens           | F0. Off Chain                                     |
```

Comment:
I'd suggest reformatting this with one status code per line, with a divider (or separate table) between ranges.

Reply:
@Arachnid 🤔 I'm not entirely certain that I follow? The intent is to show that they break cleanly into two dimensions, and meanings should be understandable when you know the categories and reasons. 

Is it because of the width of the table on GitHub? If so, I can just move the table to a website or similar.

Reply:
Once merged it'll show up on eips.ethereum.org either way. I'm just suggesting that it'd look clearer grouped by type, rather than trying to show it in a 2d matrix.

Reply:
I just noticed that you're using the low nybble for categories and the high nybble for codes in that category. That seems really weird - why not the other way around?

Reply:
At the end of the day it makes no difference to the machine, and is purely mnemonic. Totally happy to hear where the community lands on this one. This is just where our team ended up, after a few weeks of working with them.

You're probably feel some "weirdness" because it's a different scheme from HTTP. We had it the other way initially, but flipped it for a few reasons. Mostly it makes a handful of things somewhat easier. Below are the main ones. There's nothing earth shattering, but we've found it to be slightly easier, and very natural once you've used it for an hour or two.

## Short Forms

Generic is `0x0_`, general codes are consistent with their integer representations

```solidity
hex"1" == hex"01" == 1 // with casting
```

## Code Classes

Many applications will always be part of the same category. For instance, validation will generally be in the `0x10` range.

```solidity
contract Whitelist {
    mapping(address => bool) private whitelist;
    uint256 private deadline;
    byte constant private prefix = hex"10";

    check(address _, address _user) returns (byte _status) {
        if (now >= deadline)  {  return prefix | 5; } 
        if (whitelist[_user]) {  return prefix | 1; } 
        return prefix;
    }
}
```

### Helpers

This above also means that working with app-specific enums is slightly easier:

```solidity
enum Sleep {
  Awake,
  Asleep,
  REM,
  FallingAsleep
}

// From the helper library

function appCode(Sleep _state) returns (byte code) {
    return byte(160 + _state); // 160 = 0xA0
}

// Versus

function appCode(Sleep _state) returns (byte code) {
  return byte((16 * _state) + 10); // 10 = 0xA
}
```

Would love to hear thoughts from the community, though 👍🏻

Reply:
It would be worth adding that to the rationale section.